### PR TITLE
Refactor Visitor functions to allow extending it's behaviour

### PR DESCRIPTION
**Pull request by jpb**
This change moves the `visit*` functions into a class so that they can be extended for other use-cases. The main reasons I want something like this is to:

- filter the `$imports` for commands that don't need all the fields defined in the arguments file, such as the template approval commands 
  - this would remove the requirement to satisfy all imports even when they are not used
- extract the `env` `$imports` for #146 

This would also enable some nice-to-haves, such as warning when an import is not used.

- [x] remove excessive instantiation of `Visitor` in `preprocess`
- [x] see if the `VariablesVisitor` can work without having the variables defined in `env.$envValues` (that is, without having resolved the `$imports` and `$defs`)
- [x] cleanup `MyCompiler`
- [x] cleanup `visitor.ts` refactor

### CODE REVIEW

File: src/preprocess/visitor.ts [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+  $EnvValues,
+  AnyButUndefined,
+  ExtendedCfnDoc,
+  GlobalSection,
+  GlobalSectionNames,
+  MaybeStackFrame,
+  validateTemplateParameter,
+  interpolateHandlebarsString,
+  appendPath
+} from './index';
```
Comment (tavisrudd) [inline src/preprocess/visitor.ts]: Let's follow up in another PR with some refactoring to remove the circular dependency between these modules (after I've finished the other module refactoring on cfn/).

File: src/preprocess/visitor.ts [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+      } else {
+        return [`${subEnv.$envValues.Prefix}${resname}`, val];
+      }
+    });
+  }
+
+}
+
+export class VariablesVisitor extends Visitor {
+  public variables: string[] = [];
```
Comment (tavisrudd) [inline src/preprocess/visitor.ts]: Is this just for testing purposes? How about `VariableTrackingVisitor`?

**Comment by tavisrudd** [comment]
@jpb Do you want to merge #151 into this branch first prior to merging from here to master?

File: src/preprocess/visitor.ts (hunk 2 above)
Comment (jpb) [reply src/preprocess/visitor.ts]: `variables` is used in https://github.com/unbounce/iidy/blob/274f9508dd95fbb4f30c28dd321812b3369b7458/src/preprocess/filter.ts#L6-L19

I've renamed the class 👍

**Comment by tavisrudd** [comment]
Merge away

**Pull request merged**